### PR TITLE
WIP: Add support for changing cursor shape

### PR DESCRIPTION
**Pull request by jschaf**
Described in https://vt100.net/docs/vt510-rm/DECSCUSR

This is a first-attempt at implementing changing the cursor shape as described in https://stackoverflow.com/questions/29619706/intellij-terminal-cursor-shape.

Looking for feedback on the overall approach.

### CODE REVIEW

File: terminal/src/com/jediterm/terminal/ui/TerminalPanel.java [hunk 1]
```diff
@@ -970,6 +970,7 @@ public void initKeyHandler() {
     // cursor state
     private boolean myCursorIsShown; // blinking state
     protected Point myCursorCoordinates = new Point();
+    private CursorShape shape = CursorShape.BLINK_BLOCK;
```
Comment (traff) [inline terminal/src/com/jediterm/terminal/ui/TerminalPanel.java]: Could you please follow the code style and add my prefix to the field name?

File: terminal/src/com/jediterm/terminal/ui/TerminalPanel.java [hunk 2]
```diff
@@ -1119,16 +1152,6 @@ private void drawCharacters(int x, int y, TextStyle style, CharBuffer buf, Graph
       }
     }
 
-    Color backgroundColor = getPalette().getColor(myStyleState.getBackground(style.getBackgroundForRun()));
```
Comment (traff) [inline terminal/src/com/jediterm/terminal/ui/TerminalPanel.java]: Why do you remove this block of code? It clears the rectangle for a char to be printed.

**Comment by traff** [comment]
There is one more thing to consider: a way to setup a default cursor style in the terminal UI.

File: terminal/src/com/jediterm/terminal/ui/TerminalPanel.java (hunk 1 above)
Comment (Plonk42) [reply terminal/src/com/jediterm/terminal/ui/TerminalPanel.java]: It is to better show that this is an object field, and not a method variable, or a parameter, or... So you don't have to use `this.` for disambiguation and avoid confusion with local variable for instance.

**Comment by Plonk42** [comment]
Also, did you check it plays nice with "Draw cursor frame when terminal is not focused" implemented in https://github.com/JetBrains/jediterm/pull/103 ?

**Comment by traff** [comment]
Yes I mean the default cursor shape, which can be overridden by the DECSCUSR escape sequences.

**Comment by jschaf** [comment]
The default cursor shape works and is overridable by DESCSCUSR escape sequences.

The implementation plays nicely with out of focus: The behavior is as follows:

- All cursors stop blinking when terminal loses focus
- Block cursor becomes an outline instead of solid rectangle.
- Vertical line cursor remains the same.
- The underline cursor remains the same.

### Remaining items:
- figure out repaint issues when moving cursor and switching at same time.

Manual test:

**change-cursor**
```bash
#!/bin/bash
printf '\x1b['$@' q'
```

**Comment by jschaf** [comment]
From the control sequences buffer one of the problematic sequences is:

#### Change to vertical cursor; cursor forward 1
```
[6 q[1C
```

My guess is:
- initial state: cursor moves to char, inverts style for that char
- cursor changes to vertical bar
- cursor moves forward 1

It seems like the code doesn't inverse the style of the character it moved from.

**Comment by jschaf** [comment]
From the debug logs:

```
DEBUG com.jediterm.terminal.ui.AbstractTerminalFrame  - Writing in OutputStream : a
DEBUG com.jediterm.terminal.ui.AbstractTerminalFrame  - Writing in OutputStream : [97] a
DEBUG com.jediterm.terminal.emulator.JediEmulator  - Control sequence
parsed                        :ESC[6q Unhandled:  
DEBUG com.jediterm.terminal.emulator.JediEmulator  - Control sequence
parsed                        :ESC[6q
DEBUG com.jediterm.terminal.emulator.JediEmulator  - Control sequence
parsed                        :ESC[1C
```

**Comment by jschaf** [comment]
The hiding of characters exists on master so it's not a regression introduced by this PR.

**Pull request merged**